### PR TITLE
feature/cavity_noise_simulation (additive noise implementation)

### PR DESCRIPTION
**Pull request by ehtkarim**
The processor "gaussian-noise-generator" (from `KTGaussianNoiseGenerator.cc`) now correctly adds White Gaussian Noise to the signal. I've added an example config file (named `KatydidPSConfig_noise.yaml`) demonstrating the processor usage. One can also use a random number seed using the "seed" parameter in the config file

### CODE REVIEW

**Comment by ehtkarim** [comment]
- KTGaussianNoiseGenerator
   - Has a working random seed feature now
   - Can configure noise-psd-floor values. Defaults to 2.2e-13 W/Hz (default CCA scenario)
- KTCavityNoiseGenerator
   - Inherited from KTGaussianNoiseGenerator
   - Has a working random seed feature now
   - Implements Wouter/Arina's frequency dependent cavity noise model
   - Can configure model parameters in the Katydid config file
- KTMeasuredNoiseGenerator
   - Inherited from KTGaussianNoiseGenerator
   - Takes a .yaml file with psd-mean and psd-variance values at different frequencies as input, interpolates them using a 3rd order spline, and implements the psd-mean and psd-variance values per frequency bin to generate noise
   
   I'm attaching some freq. and time slices with each of the three noises applied to a nominal 90 degree **Locust_MC** signal:
   
Freq_Spectrum_PSD_Mean_1e-19_Variance_2e-36
Time_Series_PSD_Mean_1e-19_Variance_2e-36
Freq_Spectrum_cavity_retuned_parameters
Time_Series_cavity_retuned_parameters
Freq_Spectrum_aWGN_2 2e-13_PSD_floor
Time_Series_aWGN_2 2e-13_PSD_floor

File: Examples/ConfigFiles/KatydidPSNoiseConfig.yaml
Comment (nsoblath) [inline Examples/ConfigFiles/KatydidPSNoiseConfig.yaml:1]: Please rename this to `KatydidPSNoiseConfig.yaml` to stick with CamelCase and have the "noise" modifier before "Config"

File: Examples/ConfigFiles/KatydidPSNoiseCavityConfig.yaml
Comment (nsoblath) [inline Examples/ConfigFiles/KatydidPSNoiseCavityConfig.yaml:1]: Please rename this to `KatydidPSNoiseCavityConfig.yaml` to stick with CamelCase and have the "NoiseCavity" modifier before "Config"

File: Source/Simulation/KTCavityNoiseGenerator.hh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    class KTCavityNoiseGenerator : public KTGaussianNoiseGenerator
+    {
+        public:
+            KTCavityNoiseGenerator(const std::string& name = "cavity-noise-generator");
+            virtual ~KTCavityNoiseGenerator();
+
+            virtual bool ConfigureDerivedGenerator(const scarab::param_node* node);
+
+        protected:
+            struct ModelPars
```
Comment (nsoblath) [inline Source/Simulation/KTCavityNoiseGenerator.hh]: Why are the model parameters contained within a struct?  I don't see any particular design reason for it.  They should just be member variables of the processor.

I suggest making them all member variables in the class using the `MEMVAR` macros.  Try to stick with the CamelCase naming convention.

File: Source/Simulation/KTCavityNoiseGenerator.cc [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+        return true;
+    }
+
+    double KTCavityNoiseGenerator::NoisePSD(double f) const
+    {
+        const double g   = fPars.Q0 / fPars.Q_L;
+        const double lor = 1.0 / ( 1.0 + std::pow( 2.0 * fPars.Q_L * (f - fPars.f0) / fPars.f0, 2.0 ) );    // Lorentzian
+
+        const double kB  = 1.380649e-23;
```
Comment (nsoblath) [inline Source/Simulation/KTCavityNoiseGenerator.cc]: These constants should be added to KTMath (in the Utility directory).  Please add them there.  

Also, while you're editing those files, can you please make the functions that return numeric constants `constexpr`?  Originally the minimum version of C++ I used didn't have that when I first started Katydid, but we require C++17, so we an use it.

File: Source/Simulation/KTCavityNoiseGenerator.cc [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                    ts->SetValue(i, ts->GetValue(i) + noiseTS->GetReal(i));
+            }
+        }
+
+        return true;
+    }
+
+    double KTCavityNoiseGenerator::NoisePSD(double f) const
+    {
+        const double g   = fPars.Q0 / fPars.Q_L;
```
Comment (nsoblath) [inline Source/Simulation/KTCavityNoiseGenerator.cc]: I suggest you add g as a private member variable (no setter).  And then you calculate it in the setters for Q0 and Q_L.  That way you're not calculating it every time NoisePSD is calculated.

File: Source/Simulation/KTCavityNoiseGenerator.cc [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        const double omega = 2.0 * M_PI * f;
+
+        auto eta = [](double x){ return x/(std::exp(x)-1.0) + 0.5; };   // Bose-Einstein factor
+
+        const double Tcav  = fPars.T_cav  * eta(hbar*omega/(kB*fPars.T_cav));
+        const double Tisol = fPars.T_isol * eta(hbar*omega/(kB*fPars.T_isol));
+
+        const double P_cav  = kB*Tcav  * (4.*g/std::pow(1.+g,2)) * lor;
+        const double P_loss = kB*( fPars.A*fPars.T_line_start + (1.-fPars.A)*fPars.T_line_end );
+        const double P_isol = kB*Tisol * (1. - (4.*g/std::pow(1.+g,2))*lor);
```
Comment (nsoblath) [inline Source/Simulation/KTCavityNoiseGenerator.cc]: `pow()` is quite slow.  For a simple thing like this one I suggest you just use `(1.+g)*(1.+g)`.

File: Source/Simulation/KTCavityNoiseGenerator.cc [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    double KTCavityNoiseGenerator::NoisePSD(double f) const
+    {
+        const double g   = fPars.Q0 / fPars.Q_L;
+        const double lor = 1.0 / ( 1.0 + std::pow( 2.0 * fPars.Q_L * (f - fPars.f0) / fPars.f0, 2.0 ) );    // Lorentzian
+
+        const double kB  = 1.380649e-23;
+        const double hbar= 1.054571817e-34;
+        const double omega = 2.0 * M_PI * f;
+
+        auto eta = [](double x){ return x/(std::exp(x)-1.0) + 0.5; };   // Bose-Einstein factor
```
Comment (nsoblath) [inline Source/Simulation/KTCavityNoiseGenerator.cc]: I suggest implementing this as a member function.  That way you can have unit tests for it, and you won't have memory being allocated for the lambda at runtime.

File: Examples/ConfigFiles/KatydidPSNoiseConfig.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      - signal: "to-ps:psd"
+        slot: "writer:ps"
+      - signal: "egg:ts"
+        slot: "writer:ts"
+
+    run-queue:
+      - egg
+
+egg:
+    filename: "locust_mc_Seed600_Angle90.000000000_Pos0.0040000_Energy18563.251000000.egg"
```
Comment (nsoblath) [inline Examples/ConfigFiles/KatydidPSNoiseConfig.yaml]: Since this is an example, please give a generic filename.

File: Examples/ConfigFiles/KatydidPSNoiseCavityConfig.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      - signal: "to-ps:psd"
+        slot: "writer:ps"
+      - signal: "egg:ts"
+        slot: "writer:ts"
+
+    run-queue:
+      - egg
+
+egg:
+    filename: "locust_mc_Seed600_Angle90.000000000_Pos0.0040000_Energy18563.251000000.egg"
```
Comment (nsoblath) [inline Examples/ConfigFiles/KatydidPSNoiseCavityConfig.yaml]: Please give a generic filename

File: Source/Simulation/KTMeasuredNoiseGenerator.hh [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            virtual ~KTMeasuredNoiseGenerator();
+
+            virtual bool ConfigureDerivedGenerator(const scarab::param_node* node);
+
+        public:
+            virtual bool GenerateTS(KTTimeSeriesData& data);
+
+        protected:
+            // Simple natural cubic spline for doubles
+            class KCubicSpline
```
Comment (nsoblath) [inline Source/Simulation/KTMeasuredNoiseGenerator.hh]: Instead of re-implementing a spline here, I suggest using Utility/KTSpline

Comment (nsoblath) [reply Source/Simulation/KTMeasuredNoiseGenerator.hh]: Actually, you could take your implementation and replace the use of ROOT's TSpline3.  I've always been bothered by requiring ROOT for the use of KTSpline, but never enough to fix it.

**Review by nsoblath — CHANGES_REQUESTED** [review]

File: Source/Simulation/KTGaussianNoiseGenerator.cc [hunk 9]
```diff
@@ -37,16 +44,64 @@ namespace Katydid
...
+
+        input_type sigma = 0.0;
+
+        if (node->has("noise-floor-psd"))
+        {
+            sigma = std::sqrt(node->get_value<input_type>("noise-floor-psd"));
+        }
+        else if (node->has("noise-temperature"))
+        {
+            static constexpr double kBoltzmann = 1.38064852e-23; // J/K - Boltzmann constant; there's none defined in Source/Utility, so keeping a local constexpr
```
Comment (juniorpena) [inline Source/Simulation/KTGaussianNoiseGenerator.cc]: Could add kB to KTMath in Utility as Noah commented above
